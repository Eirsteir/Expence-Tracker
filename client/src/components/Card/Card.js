import React from "react";
import Paper from "@material-ui/core/Paper";
import Typography from "@material-ui/core/Typography";
import { Doughnut } from "react-chartjs-2";

class Card extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      tags: []
    };
  }

  // move all processing to Cardlist
  render() {
    const { classes, processedExpences, timespan } = this.props;
    const { tags } = this.state;

    let data = {
      labels: "",
      datasets: [
        {
          data: ""
        }
      ]
    };

    if (processedExpences.length !== 0) {
      console.time("map");
      // Filter tags so we have categories to display in the cards
      processedExpences.map((expence, i) => {
        const tag = processedExpences[i].tag;
        // add the tag and set amount to 0
        return (tags[tag] = 0);
      });

      // Add the amount to the corresponding tag
      processedExpences.map((expence, i) => {
        const tag = processedExpences[i].tag;
        const amount = processedExpences[i].amount;
        return (tags[tag] = Number(tags[tag]) + Number(amount));
      });

      // Add more colors from chartjs
      data = {
        labels: Object.keys(tags).map((tag, i) => tag),
        datasets: [
          {
            data: Object.keys(tags).map((tag, i) => this.state.tags[tag]),
            backgroundColor: [
              "#FF6384",
              "#36A2EB",
              "#FFCE56",
              "#FF6384",
              "#36A2EB",
              "#FFCE56",
              "#FF6384",
              "#36A2EB",
              "#FFCE56",
              "#FF6384",
              "#36A2EB",
              "#FFCE56"
            ],
            hoverBackgroundColor: [
              "#009688",
              "#009688",
              "#009688",
              "#009688",
              "#009688",
              "#009688",
              "#009688"
            ]
          }
        ]
      };

      console.timeEnd("map");
    }

    return (
      <Paper className="card--content">
        <h1
          style={{
            fontWeight: "300",
            padding: ".5em",
            display: "flex",
            justifyContent: "center",
            backgroundColor: "#1769aa",
            color: "#fff",
            marginTop: "0"
          }}
        >
          {timespan === "today"
            ? "Today"
            : timespan === "thisWeek"
              ? "This Week"
              : timespan === "thisMonth"
                ? "This Month"
                : "Total"}
        </h1>

        {processedExpences.length === 0 ? (
          <Typography
            variant="body1"
            component="p"
            className={classes.emptyList}
          >
            Add a new expence
          </Typography>
        ) : (
          Object.keys(tags).map((tag, i) => {
            return (
              <Typography
                variant="body1"
                component="p"
                key={i}
                className={classes.control}
              >
                {` ${tag}: ${tags[tag]}`}
              </Typography>
            );
          })
        )}
        <Doughnut data={data} />
      </Paper>
    );
  }
}

export default Card;
