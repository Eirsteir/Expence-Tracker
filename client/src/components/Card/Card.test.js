import { shallow } from 'enzyme';
import React from 'react';
import Card from './Card.js';

it('Expect to render Card component', () => {
  const mockProcessedExpencesArray = {
    allTime: [
      { timestamp: new Date(0), tag: "Food", amount: 400 }
    ]
  }

  const mockClasses = {
    root: {
      flexGrow: 1,
    },
    paper: {
      width: "17em",
      minHeight: '55vh'
    },
    control: {
      padding: '1em',
    },
    emptyList: {
      display: 'flex',
      justifyContent: 'center',
      alignItems: 'center',
      paddingTop: '2em'
    }
  }

  expect(shallow(<Card processedExpences={mockProcessedExpencesArray.allTime} timespan={mockProcessedExpencesArray[0]} classes={mockClasses}/>)).toMatchSnapshot();
})
