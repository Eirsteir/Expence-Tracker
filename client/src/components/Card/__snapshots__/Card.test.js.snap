// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expect to render Card component 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Card
    classes={
        Object {
            "control": Object {
              "padding": "1em",
            },
            "emptyList": Object {
              "alignItems": "center",
              "display": "flex",
              "justifyContent": "center",
              "paddingTop": "2em",
            },
            "paper": Object {
              "minHeight": "55vh",
              "width": "17em",
            },
            "root": Object {
              "flexGrow": 1,
            },
          }
    }
    processedExpences={
        Array [
            Object {
              "amount": 400,
              "tag": "Food",
              "timestamp": 1970-01-01T00:00:00.000Z,
            },
          ]
    }
    timespan={undefined}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <WithStyles(Typography)
          className={
                    Object {
                              "padding": "1em",
                            }
          }
          component="h3"
          style={
                    Object {
                              "backgroundColor": "#2196f3",
                              "color": "#fff",
                              "display": "flex",
                              "justifyContent": "center",
                            }
          }
          variant="headline"
>
          All Time
</WithStyles(Typography)>,
        Array [
          <WithStyles(Typography)
            className={
                        Object {
                                    "padding": "1em",
                                  }
            }
            component="p"
            variant="body1"
>
             Food: 400
</WithStyles(Typography)>,
        ],
        <Doughnut
          data={
                    Object {
                              "datasets": Array [
                                Object {
                                  "backgroundColor": Array [
                                    "#FF6384",
                                    "#36A2EB",
                                    "#FFCE56",
                                  ],
                                  "data": Array [
                                    400,
                                  ],
                                  "hoverBackgroundColor": Array [
                                    "#FF6384",
                                    "#36A2EB",
                                    "#FFCE56",
                                  ],
                                },
                              ],
                              "labels": Array [
                                "Food",
                              ],
                            }
          }
/>,
      ],
      "className": Object {
        "minHeight": "55vh",
        "width": "17em",
      },
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": "All Time",
          "className": Object {
            "padding": "1em",
          },
          "component": "h3",
          "style": Object {
            "backgroundColor": "#2196f3",
            "color": "#fff",
            "display": "flex",
            "justifyContent": "center",
          },
          "variant": "headline",
        },
        "ref": null,
        "rendered": "All Time",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "0",
        "nodeType": "class",
        "props": Object {
          "children": " Food: 400",
          "className": Object {
            "padding": "1em",
          },
          "component": "p",
          "variant": "body1",
        },
        "ref": null,
        "rendered": " Food: 400",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "data": Object {
            "datasets": Array [
              Object {
                "backgroundColor": Array [
                  "#FF6384",
                  "#36A2EB",
                  "#FFCE56",
                ],
                "data": Array [
                  400,
                ],
                "hoverBackgroundColor": Array [
                  "#FF6384",
                  "#36A2EB",
                  "#FFCE56",
                ],
              },
            ],
            "labels": Array [
              "Food",
            ],
          },
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <WithStyles(Typography)
            className={
                        Object {
                                    "padding": "1em",
                                  }
            }
            component="h3"
            style={
                        Object {
                                    "backgroundColor": "#2196f3",
                                    "color": "#fff",
                                    "display": "flex",
                                    "justifyContent": "center",
                                  }
            }
            variant="headline"
>
            All Time
</WithStyles(Typography)>,
          Array [
            <WithStyles(Typography)
              className={
                            Object {
                                          "padding": "1em",
                                        }
              }
              component="p"
              variant="body1"
>
               Food: 400
</WithStyles(Typography)>,
          ],
          <Doughnut
            data={
                        Object {
                                    "datasets": Array [
                                      Object {
                                        "backgroundColor": Array [
                                          "#FF6384",
                                          "#36A2EB",
                                          "#FFCE56",
                                        ],
                                        "data": Array [
                                          400,
                                        ],
                                        "hoverBackgroundColor": Array [
                                          "#FF6384",
                                          "#36A2EB",
                                          "#FFCE56",
                                        ],
                                      },
                                    ],
                                    "labels": Array [
                                      "Food",
                                    ],
                                  }
            }
/>,
        ],
        "className": Object {
          "minHeight": "55vh",
          "width": "17em",
        },
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "All Time",
            "className": Object {
              "padding": "1em",
            },
            "component": "h3",
            "style": Object {
              "backgroundColor": "#2196f3",
              "color": "#fff",
              "display": "flex",
              "justifyContent": "center",
            },
            "variant": "headline",
          },
          "ref": null,
          "rendered": "All Time",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "0",
          "nodeType": "class",
          "props": Object {
            "children": " Food: 400",
            "className": Object {
              "padding": "1em",
            },
            "component": "p",
            "variant": "body1",
          },
          "ref": null,
          "rendered": " Food: 400",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "data": Object {
              "datasets": Array [
                Object {
                  "backgroundColor": Array [
                    "#FF6384",
                    "#36A2EB",
                    "#FFCE56",
                  ],
                  "data": Array [
                    400,
                  ],
                  "hoverBackgroundColor": Array [
                    "#FF6384",
                    "#36A2EB",
                    "#FFCE56",
                  ],
                },
              ],
              "labels": Array [
                "Food",
              ],
            },
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
